### PR TITLE
Move gtest's main functions into separate module

### PR DESCRIPTION
**Pull request by beckerhe**
gtest's main functions used to be created by CMake on the fly. This
commit is moving them into its own cmake module and subdirectory and
treats their targets as if they were standard library targets.

It is also changing when QApplication and QCoreApplication instances are
created. By using a test listener we postpone the creation to right
before the test start.

That's important because this way gtest command line options like
`--help` or `--gtest_list_tests` won't need a QApplication instance and
hance they work in headless setups.

I discovered this problem while investigating the flaky Windows build, but
it seems unrelated to that.

### CODE REVIEW

File: src/Test/test_qtcore_main.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include <cstdio>
+#include <optional>
+
+namespace {
+// QCoreApplication objects needs to be on the stack or in static storage. Putting this variable
+// into the QCoreApplicationStarter class leads to a crash during destruction. I'm not sure what is
+// going on.
+std::optional<QCoreApplication> app;
+
+// QCoreApplicationStarter is a gtest event listener which takes are of delayed construction of the
```
Comment (danielfenner) [inline src/Test/test_qtcore_main.cpp:17]: Sentence derailed.

**Review by danielfenner — APPROVED** [review]
lgtm

File: src/Test/test_qtcore_main.cpp (hunk 1 above)
Comment (beckerhe) [reply src/Test/test_qtcore_main.cpp:17]: Oh yes. There is a `c` missing which makes it completely not understandable. Thanks. :sweat_smile:

Comment (beckerhe) [reply src/Test/test_qtcore_main.cpp]: Fixed.

**Pull request merged**